File: clients/sellingpartner-api-aa-java/tst/com/amazon/SellingPartnerAPIAA/AWSSigV4SignerTest.java
package com.amazon.SellingPartnerAPIAA;

import com.amazonaws.SignableRequest;
import com.amazonaws.auth.AWS4Signer;
import com.amazonaws.auth.AWSCredentials;
import com.amazonaws.auth.AWSCredentialsProvider;
import com.amazonaws.auth.BasicAWSCredentials;
import com.amazonaws.auth.STSAssumeRoleSessionCredentialsProvider;
import com.squareup.okhttp.Request;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.junit.MockitoJUnitRunner;

import static org.junit.Assert.assertEquals;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.doNothing;
import static org.mockito.Mockito.verify;

@RunWith(MockitoJUnitRunner.class)
public class AWSSigV4SignerTest {
    private static final String TEST_ACCESS_KEY_ID = "aKey";
    private static final String TEST_SECRET_KEY = "sKey";
    private static final String TEST_REGION = "us-east";
    private static final String TEST_ROLE_ARN = "arnrole";
    private static final String TEST_ROLESESSION_NAME = "test";

    @Mock
    private AWS4Signer mockAWS4Signer;
    @Mock
    private AWSCredentialsProvider mockAWSCredentialsProvider;
    @Mock
    private AWSCredentials mockAWSCredentials;

    private AWSSigV4Signer underTest;
    private AWSSigV4Signer underTestCredentialsProvider;

    @Before
    public void init() {
        underTest = new AWSSigV4Signer(AWSAuthenticationCredentials.builder()
                .accessKeyId(TEST_ACCESS_KEY_ID)
                .secretKey(TEST_SECRET_KEY)
                .region(TEST_REGION)
                .build()
        );
    }

    @Test
    public void signRequestUsingProvidedCredentials() {
        ArgumentCaptor<AWSCredentials> awsCredentialsArgumentCaptor = ArgumentCaptor.forClass(AWSCredentials.class);
        underTest.setAws4Signer(mockAWS4Signer);

        doNothing()
                .when(mockAWS4Signer)
                .sign(any(SignableRequest.class), awsCredentialsArgumentCaptor.capture());

        underTest.sign(new Request.Builder().url("https://api.amazon.com").build());

        AWSCredentials actualAWSCredentials = awsCredentialsArgumentCaptor.getValue();

        assertEquals(TEST_ACCESS_KEY_ID, actualAWSCredentials.getAWSAccessKeyId());
        assertEquals(TEST_SECRET_KEY, actualAWSCredentials.getAWSSecretKey());
    }

    @Test
    public void setSignerRegion() {
        assertEquals(TEST_REGION, underTest.getAws4Signer().getRegionName());
    }

    @Test
    public void setSignerServiceName() {
        assertEquals("execute-api", underTest.getAws4Signer().getServiceName());
    }

    @Test
    public void returnSignedRequest() {
        ArgumentCaptor<SignableRequest> signableRequestArgumentCaptor = ArgumentCaptor.forClass(SignableRequest.class);
        underTest.setAws4Signer(mockAWS4Signer);

        Request actualSignedRequest = underTest.sign(new Request.Builder()
                .url("http://api.amazon.com")
                .build());

        verify(mockAWS4Signer)
                .sign(signableRequestArgumentCaptor.capture(), any(AWSCredentials.class));

        SignableRequest actualSignableRequest = signableRequestArgumentCaptor.getValue();

        assertEquals(((Request)actualSignableRequest.getOriginalRequestObject()).url(), actualSignedRequest.url());
    }
    
    @Test
    public void returnSignedRequestWithCredentialProvider() {
        ArgumentCaptor<SignableRequest> signableRequestArgumentCaptor = ArgumentCaptor.forClass(SignableRequest.class);
        
        Mockito.when(mockAWSCredentialsProvider.getCredentials()).thenReturn(mockAWSCredentials);
        
        underTestCredentialsProvider = new AWSSigV4Signer(AWSAuthenticationCredentials.builder()
                .accessKeyId(TEST_ACCESS_KEY_ID)
                .secretKey(TEST_SECRET_KEY)
                .region(TEST_REGION)
                .build(), AWSAuthenticationCredentialsProvider.builder()
                .roleArn(TEST_ROLE_ARN)
                .roleSessionName(TEST_ROLESESSION_NAME)
                .build()
        );
        underTestCredentialsProvider.setAws4Signer(mockAWS4Signer);
        underTestCredentialsProvider.setAwsCredentialsProvider(mockAWSCredentialsProvider);
        
        Request actualSignedRequest = underTestCredentialsProvider.sign(new Request.Builder()
                .url("http://api.amazon.com")
                .build());

        verify(mockAWS4Signer)
                .sign(signableRequestArgumentCaptor.capture(), any(AWSCredentials.class));

        SignableRequest actualSignableRequest = signableRequestArgumentCaptor.getValue();

        assertEquals(((Request)actualSignableRequest.getOriginalRequestObject()).url(), actualSignedRequest.url());
    }
    
   
}

File: clients/sellingpartner-api-aa-php/tests/authandauth/AWSSigV4SignerTest.php
<?php

namespace SpApi\Test\AuthAndAuth;

use Aws\Credentials\AssumeRoleCredentialProvider;
use Aws\Credentials\Credentials;
use Aws\Credentials\CredentialsInterface;
use Aws\Signature\SignatureV4;
use GuzzleHttp\Promise\Promise;
use GuzzleHttp\Psr7\Request;
use PHPUnit\Framework\MockObject\MockObject;
use SpApi\AuthAndAuth\AWSAuthenticationCredentials;
use SpApi\AuthAndAuth\AWSAuthenticationCredentialsProvider;
use SpApi\AuthAndAuth\AWSSigV4Signer;
use PHPUnit\Framework\TestCase;
use Psr\Http\Message\RequestInterface;

final class AWSSigV4SignerTest extends TestCase
{
    private const TEST_ACCESS_KEY_ID = "aKey";
    private const TEST_SECRET_KEY = "sKey";
    private const TEST_REGION = "us-east";
    private const TEST_ROLE_ARN = "arnrole";
    private const TEST_SESSION_NAME = "test";

    private MockObject|SignatureV4 $mockAWS4Signer;

    private AWSSigV4Signer $underTest;

    protected function setUp(): void
    {
        $this->mockAWS4Signer = $this->createMock(SignatureV4::class);
        $this->underTest = new AWSSigV4Signer(
            new AWSAuthenticationCredentials([
                "accessKeyId" => self::TEST_ACCESS_KEY_ID,
                "secretKey" => self::TEST_SECRET_KEY,
                "region" => self::TEST_REGION
            ])
        );
    }

    public function testSignRequestUsingProvidedCredentials()
    {
        $this->underTest->setAws4Signer($this->mockAWS4Signer);

        $this->mockAWS4Signer
            ->method("signRequest")
            ->willReturnCallback(function (
                RequestInterface $requestInput,
                CredentialsInterface $credentials
            ) {
                $this->assertEquals(self::TEST_ACCESS_KEY_ID, $credentials->getAccessKeyId());
                $this->assertEquals(self::TEST_SECRET_KEY, $credentials->getSecretKey());
                return $requestInput;
            });
        $this->underTest->sign(new Request("GET", "https://api.amazon.com"));
    }

    public function testReturnSignedRequest()
    {
        $this->underTest->setAws4Signer($this->mockAWS4Signer);

        $this->mockAWS4Signer
            ->method("signRequest")
            ->willReturnCallback(function (
                RequestInterface $requestInput,
                CredentialsInterface $credentials
            ) {
                return $requestInput;
            });

        $actualSignedRequest = $this->underTest->sign(new Request("GET", "http://api.amazon.com"));

        $this->mockAWS4Signer
            ->method("signRequest")
            ->willReturnCallback(function (
                RequestInterface $requestInput,
                CredentialsInterface $credentials
            ) use ($actualSignedRequest) {
                $this->assertEquals($requestInput->getUri(), $actualSignedRequest->getUri());
                return $requestInput;
            });

        $this->underTest->sign(new Request("GET", "http://api.amazon.com"));
    }

    public function testReturnSignedRequestWithCredentialProvider()
    {
        $mockAWSCredentialsProvider = $this->createMock(AssumeRoleCredentialProvider::class);
        $mockCredentials = $this->createMock(Credentials::class);

        $promise = new Promise();
        $promise->resolve($mockCredentials);

        $underTest = new AWSSigV4Signer(
            new AWSAuthenticationCredentials([
                "accessKeyId" => self::TEST_ACCESS_KEY_ID,
                "secretKey" => self::TEST_SECRET_KEY,
                "region" => self::TEST_REGION
            ]),
            new AWSAuthenticationCredentialsProvider([
                "roleArn" => self::TEST_ROLE_ARN,
                "roleSessionName" => self::TEST_SESSION_NAME
            ])
        );

        $mockAWSCredentialsProvider
            ->method("__invoke")
            ->willReturn($promise);

        $underTest->setAws4Signer($this->mockAWS4Signer);
        $underTest->setAwsCredentialsProvider($mockAWSCredentialsProvider);

        $this->mockAWS4Signer
            ->method("signRequest")
            ->willReturnCallback(function (
                RequestInterface $requestInput,
                CredentialsInterface $credentials
            ) {
                return $requestInput;
            });

        $actualSignedRequest = $underTest->sign(new Request("GET", "http://api.amazon.com"));

        $this->mockAWS4Signer
            ->method("signRequest")
            ->willReturnCallback(function (
                RequestInterface $requestInput,
                CredentialsInterface $credentials
            ) use ($actualSignedRequest) {
                $this->assertEquals($requestInput->getUri(), $actualSignedRequest->getUri());
                return $requestInput;
            });

        $underTest->sign(new Request("GET", "http://api.amazon.com"));
    }
}

